include/seastar/core/file.hh

```diff
@@ -166,6 +172,11 @@ public:
     virtual future<> close() = 0;
     virtual std::unique_ptr<file_handle_impl> dup();
     virtual subscription<directory_entry> list_directory(std::function<future<> (directory_entry de)> next) = 0;
+#ifdef SEASTAR_COROUTINES_ENABLED
+    virtual coroutine::experimental::generator<directory_entry, dir_entry_buffer> list_directory() {
+        throw std::runtime_error("not implemented");
```

Comment:
It's a little dangerous to have a formal API return an experimental thing, no?

Reply:
It is, but the API extension thus becomes experimental itself. Those who care about stability can stick to legacy list_directory() and avoid using new list_directory() until it returns non-experimental generator.

Reply:
It's too tricky. No one will notice that it's experimental.

Reply:
how about using `[[[deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated)]]` ? i understand that this attribute has "deprecated" in its name but as the standard notes

> The [attribute-token](https://eel.is/c++draft/dcl.attr.grammar#nt:attribute-token) deprecated can be used to mark names and entities whose use is still allowed, but is discouraged for some reason[.](https://eel.is/c++draft/dcl.attr.deprecated#1.sentence-1)

see https://eel.is/c++draft/dcl.attr.deprecated. so we can add an appropriate warning message here. and once this API is proved mature enough. we can promote it and the underlying generator so they won't be marked experimental.

Reply:
Marking new API with deprecated is too tricky too :) We want the `[[experimental]]` attribute here

Reply:
Named the new method explicitly as `experimental_list_directory` in v2. Without attribute callers will need to rename something on their side anyway, be it `file fd; fd.experimental.list_directory();` or `fd.list_directory(experimental_tag{})` or `experimental(fd).list_directory();` or anything else that come to my mind :(

Reply:
> Marking new API with deprecated is too tricky too :) We want the [[experimental]] attribute here

Yet another reason to fork the compiler.

Reply:
btw this will allow us to gain experience with the interface an unmark it as experimental. We should also compare it with std::generator and see that it matches as closely as possible.